File: examples/starters/vercel-ai-sdk/starter.ts
/**
 * ACP Starter — Vercel AI SDK (TypeScript)
 *
 * The minimum code to wire ACP governance into an agent built with the
 * Vercel AI SDK (`ai` npm package, v6). Copy this folder, swap the
 * placeholder tool for your real one, ship.
 *
 * Governance pattern: decorator. `governed("tool_name", fn)` from the
 * base `@agenticcontrolplane/governance` package wraps each tool's
 * `execute` function with ACP's pre/post hooks. Vercel AI SDK calls the
 * wrapped version; denials return a tool_error string that the model
 * sees as tool output and adapts to.
 *
 * No framework-specific ACP adapter package is needed — the base
 * `governed()` helper composes cleanly with Vercel AI SDK's `tool()`.
 *
 * Run:  bash run.sh
 */

import dotenv from "dotenv";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { anthropic } from "@ai-sdk/anthropic";
import { generateText, stepCountIs, tool } from "ai";
import { z } from "zod";

import {
  configure,
  governed,
  withContext,
} from "@agenticcontrolplane/governance";

// ── 1. Load ACP credentials from ./.env next to this file.
const __dirname = path.dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: path.join(__dirname, ".env") });

const { ANTHROPIC_API_KEY, ACP_USER_TOKEN, ACP_GATEWAY_URL = "https://api.agenticcontrolplane.com" } = process.env;
if (!ANTHROPIC_API_KEY) throw new Error("Missing ANTHROPIC_API_KEY");
if (!ACP_USER_TOKEN) throw new Error("Missing ACP_USER_TOKEN (gsk_...)");

// ── 2. Point the governance SDK at your ACP gateway. Once per process.
configure({ baseUrl: ACP_GATEWAY_URL });

// ── 3. Define your tool with governance baked in. `governed(name, fn)`
// wraps your async function; Vercel AI SDK's `tool({...execute})`
// receives the wrapped version. Order and name matter — the name passed
// to `governed()` is what shows up in the ACP activity log.
// REPLACE the body with your real logic (DB lookup, API call, etc.).
const lookupRecord = governed(
  "lookup_record",
  async ({ id }: { id: string }) => {
    return { id, status: "placeholder", note: "replace me with real logic" };
  },
);

// ── 4. Agent loop via `generateText`. `stopWhen: stepCountIs(5)` caps
// the tool-use iterations (replaces the deprecated `maxSteps`).
async function run(prompt: string) {
  const result = await generateText({
    model: anthropic("claude-sonnet-4-6"),
    tools: {
      lookup_record: tool({
        description: "Look up a record by ID. Replace with your real tool description.",
        inputSchema: z.object({
          id: z.string().describe("The record ID to look up."),
        }),
        execute: lookupRecord,
      }),
    },
    stopWhen: stepCountIs(5),
    prompt,
  });
  console.log(result.text);
}

// ── 5. `withContext` binds the user's identity for every governance
// call inside. Without this wrapper, pre/post hooks silently no-op.
async function main() {
  await withContext(
    {
      userToken: ACP_USER_TOKEN!,
      agentName: "my-vercel-agent", // Rename for your agent.
      agentTier: "background",
    },
    () => run("Look up record id=abc-123 and tell me what you find."),
  );
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: examples/starters/microsoft-agent-framework/starter.py
"""
ACP Starter — Microsoft Agent Framework (Python)

The minimum code to wire ACP governance into a Microsoft Agent Framework
(MAF) agent. Copy this folder, swap the placeholder tools for your real
ones, ship.

Governance pattern: FUNCTION MIDDLEWARE. MAF has three middleware scopes
(agent run, chat, function); ACP plugs into the function scope, so ONE
middleware instance passed to `Agent(middleware=...)` governs EVERY tool
call — no per-tool decoration required. Denials set the tool result to a
tool_error string that MAF delivers to the model as tool output; the
model sees the denial and adapts.

Run:  bash run.sh
"""

from __future__ import annotations

import asyncio
import json
import os
from pathlib import Path
from typing import Any, Awaitable, Callable

from dotenv import load_dotenv
from pydantic import BaseModel

from acp_governance import configure, post_tool_output, pre_tool_use, set_context
from agent_framework import Agent, FunctionInvocationContext, FunctionMiddleware

# ── 1. Load ACP credentials from ./.env next to this file.
load_dotenv(Path(__file__).parent / ".env")


# ── 2. The seam: one FunctionMiddleware governs every tool the agent
# owns. MAF calls `process()` around each function invocation; the ACP
# pre-hook runs before `call_next()` (deny → the real function is never
# invoked), the post-hook runs after (redact/block rewrites the output).
class ACPFunctionMiddleware(FunctionMiddleware):
    """Route every MAF function/tool call through ACP's pre/post hooks."""

    async def process(
        self,
        context: FunctionInvocationContext,
        call_next: Callable[[], Awaitable[None]],
    ) -> None:
        tool_name = context.function.name
        args = context.arguments
        tool_input: dict[str, Any] = (
            args.model_dump() if isinstance(args, BaseModel) else dict(args or {})
        )

        # Pre-hook: allow/deny before the function runs.
        allowed, reason = pre_tool_use(tool_name, tool_input)
        if not allowed:
            # Do NOT call call_next(): the tool function never executes.
            # Setting context.result overrides the invocation; MAF returns
            # it to the model as the tool's output, so the model can adapt.
            context.result = f"tool_error: {reason or 'denied by policy'}"
            return

        await call_next()

        # Post-hook: audit + PII scan on the real output.
        output = context.result
        serialized = output if isinstance(output, str) else json.dumps(output, default=str)
        verdict = post_tool_output(tool_name, tool_input, serialized)
        if verdict:
            action = verdict.get("action")
            if action == "redact":
                context.result = verdict.get("modified_output", "[ACP] Redacted")
            elif action == "block":
                context.result = f"[ACP] Blocked: {verdict.get('reason', 'policy')}"


# ── 3. Your tools. Plain Python functions — MAF derives each tool's
# schema and description from the signature, type hints, and docstring.
# No ACP decoration needed: the middleware above covers all of them.
# REPLACE the bodies with your real logic (DB lookup, API call, etc.).
def lookup_record(id: str) -> str:
    """Look up a record by ID. Replace with your real tool description."""
    return json.dumps({"id": id, "status": "placeholder", "note": "replace me"})


def send_email(to: str, subject: str, body: str) -> str:
    """Send an email. Replace with your real mailer."""
    return f"queued email to {to!r} subject={subject!r} ({len(body)} chars)"


async def main() -> None:
    for v in ("ANTHROPIC_API_KEY", "ACP_USER_TOKEN"):
        if not os.environ.get(v):
            raise SystemExit(f"Missing {v} — copy .env.example → .env")

    # ── 4. Point the governance SDK at your ACP gateway. Once per process.
    configure(
        base_url=os.environ.get("ACP_GATEWAY_URL", "https://api.agenticcontrolplane.com"),
        client_header="acp-maf-starter/0.1.0",
    )

    # ── 5. Bind identity for every governance call inside. Without this
    # set_context call, ACP pre/post hooks silently no-op.
    set_context(
        user_token=os.environ["ACP_USER_TOKEN"],
        agent_name="my-maf-agent",  # Rename for your agent.
        agent_tier="background",
    )

    # ── 6. Build the agent. Swap AnthropicClient for any MAF chat client
    # (agent_framework.openai.OpenAIChatClient, agent_framework.azure, ...)
    # — the middleware is client-agnostic.
    from agent_framework.anthropic import AnthropicClient

    agent = Agent(
        client=AnthropicClient(model_id="claude-sonnet-4-6"),
        instructions="You are an ACP-governed agent. Use the tools available.",
        name="my-maf-agent",
        tools=[lookup_record, send_email],
        middleware=ACPFunctionMiddleware(),
    )

    response = await agent.run("Look up record id=abc-123 and tell me what you find.")
    print(response.text)


if __name__ == "__main__":
    asyncio.run(main())
